rust/src/reactive/flat_fat.rs

```diff
@@ -0,0 +1,185 @@
+
+    /// Creates a new FAT with uninitialized values
+    fn with_capacity(capacity: usize) -> Self;
+
+    /// Updates a non-contiguous batch of leaves
+    fn update<'a, I>(&mut self, batch: I)
+    where
+        I: IntoIterator<Item = &'a (usize, Value)>,
+        Value: 'a;
+
+    /// Updates a contiguous batch of leaves
+    fn update_ordered<'a, I>(&mut self, values: I)
+    where
+        I: IntoIterator<Item = &'a Value>,
+        Value: 'a;
+
+    /// Aggregates all nodes in the FAT and returns the result
+    fn aggregate(&self) -> Value;
+
+    /// Aggregates a prefix of nodes in the FAT and returns the result
+    fn prefix(&self, i: usize) -> Value;
+
+    /// Aggregates a suffix of nodes in the FAT and returns the result
+    fn suffix(&self, i: usize) -> Value;
+}
+
+pub struct FlatFAT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    /// A flat binary tree, indexed as:
+    ///       0
+    ///      / \
+    ///     /   \
+    ///    1     2
+    ///   / \   / \
+    ///  3   4 5   6
+    pub(crate) tree: Vec<Value>,
+    /// Number of leaves which can be stored in the tree
+    pub(crate) capacity: usize,
+    binop: std::marker::PhantomData<BinOp>,
+}
+
+impl<Value, BinOp> FlatFAT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    /// Returns all leaf nodes of the tree
+    pub(crate) fn leaves(&self) -> &[Value] {
+        &self.tree[self.leaf(0)..]
+    }
+    /// Returns the index of the root node
+    fn root(&self) -> usize {
+        0
+    }
+    /// Returns the index of a leaf node
+    fn leaf(&self, i: usize) -> usize {
+        i + self.capacity - 1
+    }
+    /// Returns the index of an node's left child
+    fn left(&self, i: usize) -> usize {
+        2 * (i + 1) - 1
+    }
+    /// Returns the index of an node's right child
+    fn right(&self, i: usize) -> usize {
+        2 * (i + 1)
+    }
+    /// Returns the index of an node's parent
+    fn parent(&self, i: usize) -> usize {
+        (i - 1) / 2
+    }
+}
+
+impl<Value, BinOp> FAT<Value, BinOp> for FlatFAT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    fn new(values: &[Value]) -> Self {
+        let capacity = values.len();
+        let mut new = Self::with_capacity(capacity);
+        new.update_ordered(values);
+        new
+    }
+
+    fn with_capacity(capacity: usize) -> Self {
+        assert_ne!(capacity, 0, "Capacity of window must be greater than 0");
+        Self {
+            tree: vec![Value::identity(); 2 * capacity - 1],
+            binop: std::marker::PhantomData,
+            capacity,
+        }
+    }
+
+    fn update<'a, I>(&mut self, batch: I)
+    where
+        I: IntoIterator<Item = &'a (usize, Value)>,
+        Value: 'a,
+    {
+        let mut parents: HashSet<usize> = batch
+            .into_iter()
+            .map(|(idx, val)| {
+                let leaf = self.leaf(*idx);
```

Comment:
Forgive me, I'm still new to Rust, and I'm trying to really understand all of the code: why is `idx` a reference here?

Reply:
No problem, you have a good point, it probably does not need to be a reference here. The reason I had it originally was that passed an array of values by reference into the update function. When you iterate over an array (e.g., `[10; Value]`or slice (`&[Value]`) with `.iter()`, the items are by-reference. It is possible to iterate by-value over arrays in nightly Rust (https://doc.rust-lang.org/std/array/struct.IntoIter.html). From what I know, the solution in stable Rust is to iterate by-reference and clone the items into values.